fixbackend/auth/router.py

```diff
@@ -122,14 +129,94 @@ def auth_router(config: Config, google_client: GoogleOAuth2, github_client: Gith
-    )
+    @router.post("/mfa/add")
+    async def add_mfa(user: AuthenticatedUser, user_manager: UserManager = Depends(get_user_manager)) -> OTPConfig:
+        if user.is_mfa_active:
+            raise HTTPException(
+                status_code=status.HTTP_400_BAD_REQUEST,
+                detail="MFA already enabled",
+            )
+        # create a new one-time password
+        user_secret = pyotp.random_base32()
+        totp = pyotp.totp.TOTP(user_secret)
+        # store the secret
+        await user_manager.user_repository.update(user, {"otp_secret": user_secret, "is_mfa_active": False})
+        # return the OTP Config
+        return OTPConfig(uri=totp.provisioning_uri(name=user.email, issuer_name="Fix", image=OTP_Logo_URL))
+
+    @router.post("/mfa/enable")
+    async def enable_mfa(
+        user: AuthenticatedUser, otp: str = Form(), user_manager: UserManager = Depends(get_user_manager)
+    ) -> Response:
+        if user.is_mfa_active:
+            raise HTTPException(
+                status_code=status.HTTP_400_BAD_REQUEST,
+                detail="MFA already enabled",
+            )
+        if (secret := user.otp_secret) is None or not pyotp.TOTP(secret).verify(otp):
+            raise HTTPException(
+                status_code=status.HTTP_428_PRECONDITION_REQUIRED,
+                detail="MFA_NOT_PROVIDED_OR_INVALID",
+            )
+        await user_manager.user_repository.update(user, {"is_mfa_active": True})
+        return Response(status_code=status.HTTP_204_NO_CONTENT)
+
+    @router.post("/mfa/disable")
+    async def disable_mfa(
+        user: AuthenticatedUser,
+        otp: Optional[str] = Form(default=None, description="One time password"),
+        recovery_code: Optional[str] = Form(default=None, description="Recovery code"),
+        user_manager: UserManager = Depends(get_user_manager),
+    ) -> Response:
+        if user.is_mfa_active:
+            if (otp is None and recovery_code is None) or (
+                (otp_defined := otp)
+                and (secret := user.otp_secret) is not None
+                and not pyotp.TOTP(secret).verify(otp_defined)
+            ):
+                raise HTTPException(
+                    status_code=status.HTTP_428_PRECONDITION_REQUIRED,
+                    detail="MFA_NOT_PROVIDED_OR_INVALID",
+                )
+            await user_manager.user_repository.update(user, {"is_mfa_active": False, "otp_secret": None})
+        return Response(status_code=status.HTTP_204_NO_CONTENT)
+
+    # TODO: remove jwt from path
+    @router.post("/jwt/login", name=f"auth:{auth_backend.name}.login")
+    async def login(
+        request: Request,
+        credentials: OAuth2PasswordMFARequestForm = Depends(),
+        user_manager: UserManager = Depends(get_user_manager),
+        strategy: FixJWTStrategy = Depends(auth_backend.get_strategy),
+    ) -> Response:
```

Comment:
Shouldn't we have a 2-step auth process here? Otherwise our FE has no idea if it should ask for the TOTP code for any particular user.

Reply:
It is a 2 step process. MFA enabled logins will return a 428 in case of missing otp.